sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -294,7 +296,12 @@ object ShuffleExchangeExec {
           sorter.sort(iter.asInstanceOf[Iterator[UnsafeRow]])
         }
       } else {
-        rdd
+        part match {
+          case partitioner: RangePartitioner[InternalRow @unchecked, _]
+            if partitioner.getSampledArray != null =>
+            sparkContext.parallelize(partitioner.getSampledArray.toSeq, rdd.getNumPartitions)
```

Comment:
When you just parallelize sampled data, it might not have required partitioning (rang partitioning), doesn't it affect the result?

Reply:
The parallelized sampled data will be assigned to the newRdd.  the newRdd will execute:
```
newRdd.mapPartitionsInternal { iter =>
          val getPartitionKey = getPartitionKeyExtractor()
          val mutablePair = new MutablePair[Int, InternalRow]()
          iter.map { row => mutablePair.update(part.getPartition(getPartitionKey(row)), row) }
```
the newRdd uses the partitioner actually. It will map row to (partitionId, row) for the further shuffle